File: algorithms/go/search-in-rotated-sorted-array/search-in-rotated-sorted-array.go
func search(nums []int, target int) int {
	var left, right int = 0, len(nums) - 1
	for left < right {
		var mid int = (left + right) / 2
		if nums[mid] < nums[right] {
			right = mid
		} else {
			left = mid + 1
		}
	}
	var rot = left
	left, right = 0, len(nums)-1
	for left <= right {
		var mid = (left + right) / 2
		var realMid = (mid + rot) % len(nums)
		if nums[realMid] == target {
			return realMid
		} else if nums[realMid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return -1
}

File: algorithms/javascript/search-in-rotated-sorted-array/search-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
/*
    https://leetcode.com/problems/search-in-rotated-sorted-array/discuss/273622/Javascript-Simple-O(log-N)-Binary-Search-Solution

    另一种解法，rotated数组，分成一半，一定有一半是有序的，那么不断的找有序的这一半，然后再找target
*/
 var search = function(nums, target) {
    let left = 0;
    let right = nums.length - 1;
      
    while (left <= right) {
      let mid = Math.floor((left + right) / 2);
      
      if (nums[mid] === target) {
        return mid;
      }
      
      // When dividing the roated array into two halves, one must be sorted.
      
      // Check if the left side is sorted
      if (nums[left] <= nums[mid]) {
        if (nums[left] <= target && target <= nums[mid]) {
          // target is in the left
          right = mid - 1;
          
        } else {
          // target is in the right
          left = mid + 1;
        }
      } 
      
      // Otherwise, the right side is sorted
      else {
        if (nums[mid] <= target && target <= nums[right]) {
          // target is in the right
          left = mid + 1;
  
        } else {
          // target is in the left
          right = mid - 1;
        }
      }
      
      
    }
      
    return -1;
};